// @astrojs/parser/dist/parse/read/style.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var style_exports = {};
__export(style_exports, {
  default: () => read_style
});
function read_style(parser, start, attributes) {
  const content_start = parser.index;
  const styles = parser.read_until(/<\/style>/);
  const content_end = parser.index;
  parser.eat("</style>", true);
  const end = parser.index;
  return {
    type: "Style",
    start,
    end,
    attributes,
    content: {
      start: content_start,
      end: content_end,
      styles
    }
  };
}
function is_ref_selector(a, b) {
  if (!b)
    return false;
  return a.type === "TypeSelector" && a.name === "ref" && b.type === "PseudoClassSelector";
}
module.exports = __toCommonJS(style_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
