// @astrojs/parser/dist/utils/error.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var error_exports = {};
__export(error_exports, {
  CompileError: () => CompileError,
  default: () => error
});
var import_locate_character = require("locate-character");
var import_get_code_frame = __toESM(require("./get_code_frame.js"));
class CompileError extends Error {
  constructor({ code, filename, start, end, message }) {
    super(message);
    this.start = (0, import_locate_character.locate)(code, start, { offsetLine: 1 });
    this.end = (0, import_locate_character.locate)(code, end || start, { offsetLine: 1 });
    this.filename = filename;
    this.message = message;
    this.frame = (0, import_get_code_frame.default)(code, this.start.line - 1, this.start.column);
  }
  toString() {
    return `${this.filename}:${this.start.line}:${this.start.column}
	${this.message}
${this.frame}`;
  }
}
function error(code, message, props) {
  const err = new CompileError({ code, message, start: props.start, end: props.end, filename: props.filename });
  err.name = props.name;
  throw err;
}
module.exports = __toCommonJS(error_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CompileError
});
