// @astrojs/parser/dist/utils/namespaces.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var namespaces_exports = {};
__export(namespaces_exports, {
  foreign: () => foreign,
  html: () => html,
  mathml: () => mathml,
  namespaces: () => namespaces,
  svg: () => svg,
  valid_namespaces: () => valid_namespaces,
  xlink: () => xlink,
  xml: () => xml,
  xmlns: () => xmlns
});
const foreign = "https://svelte.dev/docs#svelte_options";
const html = "http://www.w3.org/1999/xhtml";
const mathml = "http://www.w3.org/1998/Math/MathML";
const svg = "http://www.w3.org/2000/svg";
const xlink = "http://www.w3.org/1999/xlink";
const xml = "http://www.w3.org/XML/1998/namespace";
const xmlns = "http://www.w3.org/2000/xmlns";
const valid_namespaces = ["foreign", "html", "mathml", "svg", "xlink", "xml", "xmlns", foreign, html, mathml, svg, xlink, xml, xmlns];
const namespaces = { foreign, html, mathml, svg, xlink, xml, xmlns };
module.exports = __toCommonJS(namespaces_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  foreign,
  html,
  mathml,
  namespaces,
  svg,
  valid_namespaces,
  xlink,
  xml,
  xmlns
});
